File: tracy/src/mathlib.cc
/* Tracy-2

   J. Bengtsson, CBP, LBL      1990 - 1994   Pascal version
                 SLS, PSI      1995 - 1997
   M. Boege      SLS, PSI      1998          C translation
   L. Nadolski   SOLEIL        2002          Link to NAFF, Radia field maps
   J. Bengtsson  NSLS-II, BNL  2004 -        

*/

// missing in lstdc++
//template double std::__cmath_power<double>(double, unsigned);

double log(const int k) { return log((double)k); }


/* Local variables for DetMat: */
struct LOC_DetMat
{
  const Matrix  *a;
  bool    cross[ss_dim];
};

typedef int iv1[ss_dim];
typedef int iv2[ss_dim][2];
typedef double v1[ss_dim];

/* Local variables for InvMat: */
struct LOC_InvMat
{
  long    n;
  Matrix  *a;
  long    row, column;
  double  determ;
} ;

void iniranf(const long i)
{
  rseed0 = i; rseed = i;
}

#define k               19
#define c               656329L
#define m               100000001
void newseed(void)
{

  rseed0 = (k*rseed0+c) % m; rseed = (rseed0+54321) % m;
}

double ranf(void)
{
  /* Generate random number with rectangular distribution */
  rseed = (k*rseed+c) % m; return (rseed/1e8);
}

#undef k
#undef c
#undef m


void setrancut(const double cut)
{

  printf("\n");
  printf("setrancut: cut set to %3.1f\n", cut);

  normcut_ = cut;
}


#define maxiter         100 
double normranf(void)
{  
  int i, j;
  double f, w;

  j = 0;
  do {
    j++;
    w = 0.0;
    for (i = 1; i <= 12; i++)
      w += ranf();
    f = w - 6.0;
  }
  while (fabs(f) > fabs(normcut_) && j <= maxiter);

  if (j > maxiter)
    fprintf(stdout,"  *** fatal error in normranf\n");
  return f;
}
#undef maxiter


double dtor(const double d) { return (d*M_PI/180.0); }


double GetAngle(const double x, const double y)
{
  double z;
  
//  if (pi == 0e0)
//    fprintf(stdout,"** pi not initialized in GetAngle\n");
  if (x != 0e0)
    z = atan(y/x);
  else
    z = sgn(y)*M_PI/2.0;
  if (x >= 0.0)
    return z;
  if (y >= 0.0)
    z += M_PI;
  else
    z -= M_PI;
  return z;
}


void CopyVec(const int n, const Vector &a, Vector &b)
{
  int i;

  for (i = 0; i < n; i++)
    b[i] = a[i];
}


void AddVec(const int n, const Vector &a, Vector &b)
{
  int i;
  for (i = 0; i < n; i++)
    b[i] = a[i] + b[i];
}


void SubVec(int n, const Vector &a, Vector &b)
{
  int i;

  for (i = 0; i < n; i++)
    b[i] -= a[i];
}


double xabs(long n, Vector &x)
{
  long    i;
  double  sum;

  sum = 0.0;
  for (i = 0; i < n; i++)
    sum += sqr(x[i]);

  return sqrt(sum);
}


void UnitMat(const int n, Matrix &A)
{
  int i, j;

  for (i = 1; i <= n; i++) {
    for (j = 1; j <= n; j++) {
      if (i == j)
        A[i-1][j-1] = 1.0;
      else
        A[i-1][j-1] = 0.0;
    }
  }
}


void ZeroMat(const int n, Matrix &A)
{
  int i, j;

  for (i = 0; i < n; i++) {
    for (j = 0; j < n; j++)
      A[i][j] = 0.0;
  }
}


void CopyMat(const int n, const Matrix &A, Matrix &B)
{
  int i;

  for (i = 0; i < n; i++)
    CopyVec(n, A[i], B[i]);
}


void AddMat(const int n, const Matrix &a, Matrix &b)
{
  int i;

  for (i = 0; i < n; i++)
    AddVec(n, a[i], b[i]);
}

void SubMat(const int n, const Matrix &a, Matrix &b)
{
  /*n : integer; VAR a, b : matrix*/
  int i;

  for (i = 0; i < n; i++)
    SubVec(n, a[i], b[i]);
}


void LinTrans(const int n, const Matrix &a, Vector &x)
{
  int i, j;
  Vector y;

  for (i = 0; i < n; i++) {
    y[i] = 0e0;
    for (j = 0; j < n; j++)
      y[i] += a[i][j]*x[j];
  }
  CopyVec(n, y, x);
}


void MulcMat(const int n, const double c, Matrix &A)
{
  int i,j;

  for (i = 0; i < n; i++)
    for (j = 0; j < n; j++)
      A[i][j] = A[i][j]*c;
}


void MulLMat(const int n, const Matrix &A, Matrix &B)
{
  int i, j, k;
  double x;
  Matrix C;

  for (i = 0; i < n; i++)
    for (j = 0; j < n; j++) {
      x = 0e0;
      for (k = 0; k < n; k++)
        x += A[i][k]*B[k][j];
      C[i][j] = x;
    }

  CopyMat(n, C, B);
}


void MulRMat(const int n, Matrix &A, const Matrix &B)
{
  int i, j, k;
  double x;
  Matrix C;

  for (i = 0; i < n; i++)
    for (j = 0; j < n; j++) {
      x = 0e0;
      for (k = 0; k < n; k++)
        x += A[i][k]*B[k][j];
      C[i][j] = x;
    }

  CopyMat(n, C, A);
}


double TrMat(const int n, const Matrix &A)
{
  int i;
  double x;

  x = 0e0;
  for (i = 0; i < n; i++)
    x += A[i][i];
  return x;
}


void TpMat(const int n, Matrix &A)
{
  int i, j;
  double x;

  for (i = 1; i <= n; i++) {
    for (j = 0; j <= i-2; j++) {
      x = A[i-1][j];
      A[i-1][j] = A[j][i-1];
      A[j][i-1] = x;
    }
  }
}


void SwapSigmaMat(Matrix &A)
{
  int i, j;
  double x;

  for (i = 0; i <= 2; i++) {
    for (j = 0; j <= 2; j++) {
      x = A[2*i][2*j]; A[2*i][2*j] = A[2*i+1][2*j+1]; A[2*i+1][2*j+1] = x;
      x = A[2*i][2*j+1]; A[2*i][2*j+1] = A[2*i+1][2*j]; A[2*i+1][2*j] = x;
    }
  }
}


double GdetMat(long n, struct LOC_DetMat *LINK)
{
  double Result = 0.0;
  long k, sign;
  double det;

  if (n > 1)
  {
    det = 0e0;
    if ((n & 1) == 1)
      sign = 1;
    else
      sign = -1;
    for (k = 0; k < ss_dim; k++) {
      if (!LINK->cross[k]) {
        LINK->cross[k] = true;
        det += sign * (*LINK->a)[n - 1][k] * GdetMat(n - 1, LINK);
        LINK->cross[k] = false;
        sign = -sign;
      }
    }
    return det;
  }
  for (k = 0; k < ss_dim; k++) {
    if (!LINK->cross[k])
      Result = (*LINK->a)[n - 1][k];
  }
  return Result;
}


double DetMat(const int n, const Matrix &A_)
{
  struct LOC_DetMat V;
  long j;
  double d;

  V.a = &A_;
  if (n == 2)  /* simple case of a matrix of rank 2*/
    return ((*V.a)[0][0] * (*V.a)[1][1] - (*V.a)[0][1] * (*V.a)[1][0]);
  else if (n == 3) { /* simple case of a matrix of rank 3*/
    d = (*V.a)[0][0]
        * ((*V.a)[1][1] * (*V.a)[2][2] - (*V.a)[1][2] * (*V.a)[2][1]);
    d += (*V.a)[0][1]
         * ((*V.a)[1][2] * (*V.a)[2][0] - (*V.a)[1][0] * (*V.a)[2][2]);
    d += (*V.a)[0][2]
         * ((*V.a)[1][0] * (*V.a)[2][1] - (*V.a)[1][1] * (*V.a)[2][0]);
    return d;
  } else {
    for (j = 1; j <= ss_dim; j++) {
      if (j <= n)
        V.cross[j - 1] = false;
      else
        V.cross[j - 1] = true;
    }
    return (GdetMat(n, &V));
  }
}


void swap_(double *x, double *y, struct LOC_InvMat *LINK)
{
  double d;

  d  = *x;
  *x = *y;
  *y = d;
}

void Interchange(struct LOC_InvMat *LINK)
{
  long l, FORLIM;

  if (LINK->row == LINK->column)
    return;
  LINK->determ = -LINK->determ;
  FORLIM = LINK->n;
  for (l = 0; l < FORLIM; l++)
    swap_(&(*LINK->a)[LINK->row - 1][l],
	  &(*LINK->a)[LINK->column - 1][l], LINK);
}


bool InvMat(const int n_, Matrix &A_)
{
  struct LOC_InvMat V;
  bool Result = false;
  long i, j, k, l, l1;
  double amax, t, d;
  Matrix b;
  iv1 ipivot;
  iv2 index;
  v1 pivot;
  long FORLIM, FORLIM1;

  V.n = n_;
  V.a = &A_;

  /* if 2-square matrix */
  if (V.n == 2) {
    d = DetMat(2, *V.a);
    if (d != 0e0) {  /* non zero determinant */
      Result = true;
      b[0][0] = (*V.a)[1][1] / d;
      b[0][1] = -((*V.a)[0][1] / d);
      b[1][0] = -((*V.a)[1][0] / d);
      b[1][1] = (*V.a)[0][0] / d;
      CopyMat(V.n, b, *V.a);
    } else /* non iversible matrix */
      Result = false;
  } else { /* matrix with n greater than 2 */
    V.determ = 1.0;
    FORLIM = V.n;
    for (j = 0; j < FORLIM; j++)
      ipivot[j] = 0;
    i = 1;
    while (i <= V.n && V.determ != 0e0) {
      amax = 0e0;
      FORLIM = V.n;
      for (j = 1; j <= FORLIM; j++) {
        if (ipivot[j - 1] != 1) {
          FORLIM1 = V.n;
          for (k = 1; k <= FORLIM1; k++) {
            if (ipivot[k - 1] > 1) goto _L1;
            if (ipivot[k - 1] < 1) {
              if (fabs(amax) < fabs((*V.a)[j - 1][k - 1])) {
                V.row = j;
                V.column = k;
                amax = (*V.a)[j - 1][k - 1];
              }
            }
          }
        }
      }
      if (amax == 0e0) {
        Result = false;
        V.determ = 0e0;
      } else {
        Result = true;
        ipivot[V.column - 1]++;
        Interchange(&V);
        index[i - 1][0] = V.row;
        index[i - 1][1] = V.column;
        pivot[i - 1] = (*V.a)[V.column - 1][V.column - 1];
        V.determ *= pivot[i - 1];
        (*V.a)[V.column - 1][V.column - 1] = 1.0;
        FORLIM = V.n;
        for (l = 0; l < FORLIM; l++)
          (*V.a)[V.column - 1][l] /= pivot[i - 1];
        FORLIM = V.n;
        for (l1 = 0; l1 < FORLIM; l1++) {
          if (l1 + 1 != V.column) {
            t = (*V.a)[l1][V.column - 1];
            (*V.a)[l1][V.column - 1] = 0e0;
            FORLIM1 = V.n;
            for (l = 0; l < FORLIM1; l++)
              (*V.a)[l1][l] -= (*V.a)[V.column - 1][l] * t;
          }
        }
      }  /*else */
      i++;
    }  /*while*/
    if (V.determ != 0e0) {
      FORLIM = V.n;
      for (i = 1; i <= FORLIM; i++) {
        l = V.n - i + 1;
        if (index[l - 1][0] != index[l - 1][1]) {
          V.row = index[l - 1][0];
          V.column = index[l - 1][1];
          FORLIM1 = V.n;
          for (k = 0; k < FORLIM1; k++)
            swap_(&(*V.a)[k][V.row - 1], &(*V.a)[k][V.column - 1], &V);
        }
      }
    }
  }
_L1:
  return Result;
}


#define SWAP(a,b) {float temp=(a);(a)=(b);(b)=temp;}

bool InvMat2(double a[4][4])
{
  const int n = 4;

  int     indxc[n], indxr[n], ipiv[n];
  int     i, icol = 0, irow = 0, j, k, l, ll;
  double  big, dum, pivinv;

  for (j=0;j<n;j++)
    ipiv[j]=0;
  for (i=0;i<n;i++) {
    big=0.0;
    for (j=0;j<n;j++)
      if (ipiv[j] != 1)
	for (k=0;k<n;k++) {
	  if (ipiv[k] == 0) {
	    if (fabs(a[j][k]) >= big) {
	      big=fabs(a[j][k]);
	      irow=j;
	      icol=k;
	    }
	  } else if (ipiv[k] > 1) fprintf(stdout,"GAUSSJ: Singular Matrix-1");
				}
    ++(ipiv[icol]);
    if (irow != icol) {
      for (l=0;l<n;l++) SWAP(a[irow][l],a[icol][l])
			   //~ for (l=0;l<=n;l++) SWAP(b[irow][l],b[icol][l])
			   }
    indxr[i]=irow;
    indxc[i]=icol;
    if (a[icol][icol] == 0.0) fprintf(stdout,"GAUSSJ: Singular Matrix-2");
    pivinv=1.0/a[icol][icol];
    a[icol][icol]=1.0;
    for (l=0;l<n;l++) a[icol][l] *= pivinv;
    //~ for (l=0;l<n;l++) b[icol][l] *= pivinv;
    for (ll=0;ll<n;ll++)
      if (ll != icol) {
	dum=a[ll][icol];
	a[ll][icol]=0.0;
	for (l=0;l<n;l++) a[ll][l] -= a[icol][l]*dum;
	//~ for (l=0;l<n;l++) b[ll][l] -= b[icol][l]*dum;
      }
  }
  for (l=n-1;l>=0;l--) {
    if (indxr[l] != indxc[l])
      for (k=0;k<n;k++)
	SWAP(a[k][indxr[l]],a[k][indxc[l]]);
  }
  return true;
}
#undef SWAP


void prtmat(const int n, const Matrix &A)
{
  int i, j;

  printf("matrix:\n");
  for (i = 0; i < n; i++) {
    for (j = 0; j < n; j++)
      printf(" %14.6e", A[i][j]);
//      printf(" %24.16e", A[i][j]);
    putchar('\n');
  }
}

File: Tracy-3.5_macOS/tracy-2_pas/src/mathlib.pas
  module mathlib(input, output);

  %include 'mathlibo.def'

  { Interface }

  var	pi		: [global] double;
	rseed0, rseed	: [global] integer;
	normcut_	: [global] double;

  { Note that pi has to be initialized }

  { Extensions to the standard functions }

{  [global] FUNCTION dble(x : real) : double; forward;}

{  [global] FUNCTION sngl(x : double) : real; forward;}

  [global] FUNCTION min_(x1, x2 : double) : double; forward;

  [global] FUNCTION max_(x1, x2 : double) : double; forward;

  [global] FUNCTION pwr(x, y : double) : double; forward;
  { power <- x^y }

  [global] FUNCTION tan_(x : double) : double; forward;

  [global] FUNCTION cosh_(x : double) : double; forward;

  [global] FUNCTION sinh_(x : double) : double; forward;

  [global] FUNCTION tanh_(x : double) : double; forward;

  [global] PROCEDURE iniranf(i : integer); forward;

  [global] PROCEDURE newseed; forward;

  [global] FUNCTION ranf : double; forward;

  [global] PROCEDURE setrancut(cut : double); forward;

  [global] FUNCTION normranf : double; forward;

  { Conversion routines }

  [global] FUNCTION dtor (d : double) : double; forward;

  [global] FUNCTION sign(x : double) : integer; forward;

  [global] Function GetAngle(x, y : double) : double; forward;
  { Get the angle phi from x=cos(phi), y=sin(phi) where -pi <= phi <= pi }

  { Matrix routines }

  [global] PROCEDURE UnitMat(n : integer; VAR a : matrix); forward; 

  [global] PROCEDURE CopyVec(n : integer; VAR a, b : vector); forward;

  [global] PROCEDURE CopyMat(n : integer; VAR a, b : matrix); forward;

  [global] PROCEDURE AddVec(n : integer; VAR a, b : vector); forward;

  [global] PROCEDURE SubVec(n : integer; VAR a, b : vector); forward;

  [global] PROCEDURE AddMat(n : integer; VAR a, b : matrix); forward;

  [global] PROCEDURE SubMat(n : integer; VAR a, b : matrix); forward;

  [global] PROCEDURE LinTrans(n : integer; VAR a : matrix; VAR x : vector); forward;

  [global] PROCEDURE MulLMat(n : integer; VAR a, b : matrix); forward;

  [global] PROCEDURE MulRMat(n : integer; VAR a, b : matrix); forward;

  [global] FUNCTION TrMat(n : integer; VAR a : matrix) : double; forward;

  [global] PROCEDURE TpMat(n : integer; VAR a : matrix); forward;

  [global] FUNCTION DetMat(n : integer; VAR a : matrix) : double; forward;

  [global] function InvMat(n : integer; VAR a : matrix) : boolean; forward;

  [global] procedure prtmat(n : integer; var a : matrix); forward;

  { Implementation }

{  FUNCTION sngl}{x : double}{ : real}{;

  begin
    sngl := x;
  end;}


{  FUNCTION dble}{x : real}{ : double}{;

  begin
    dble := x;
  end;}


  FUNCTION min_{x1, x2 : double}{ : double};

  begin
    if x1 < x2 then
      min_ := x1
    else
      min_ := x2;
  end;


  FUNCTION max_{x1, x2 : double}{ : double};

  begin
    if x1 > x2 then
      max_ := x1
    else
      max_ := x2;
  end;


  function pwr{x, y : double) : double};

  { pwr <- x^y  }

  begin
    if x = 0d0 then
    begin
      if y = 0d0 then
      begin
	pwr := 1d0;
	writeln('  *** 0^0 = 1');
      end
      else
	pwr := 0d0;
    end
    else if x > 0d0 then
      pwr := exp(y*ln(x))
    else
    begin
      if y = round(y) then
      begin
	if not odd(round(y)) then
          pwr := exp(y*ln(-x))
	else
          pwr := -exp(y*ln(-x))
      end
      else
	writeln('  *** undefined exponentiation');
    end;
  end;


  FUNCTION tan_{x : double}{ : double};

  BEGIN
    tan_:=sin(x)/cos(x)
  END;


  FUNCTION cosh_{x : double}{ : double};

  VAR  y : double  ;

  BEGIN
    y := exp(x); cosh_ := (y + 1/y)/2;
  END;


  FUNCTION sinh_{x : double}{ : double};

  VAR  y : double  ;

  BEGIN
    y := exp(x); sinh_ := (y - 1/y)/2
  END;


  FUNCTION tanh_{x : double}{ : double};

  VAR  y, z : double  ;

  BEGIN
    y := exp(x); z := 1/y; tanh_ := (y - z)/(y + z)
  END;


  PROCEDURE iniranf{i : integer};

  { Initialize random number generator with seed i }

  BEGIN 
     rseed0 := i; rseed := i;
  END;


  PROCEDURE newseed;

  { Get new seed }

  const k=19; c=656329; m=100000001;

  BEGIN 
     rseed0 := ( k*rseed0 + c ) mod m;
     rseed := ( rseed0 + 54321 ) mod m;
  END;


  FUNCTION ranf {: double};

  { Generate random number with rectangular distribution }

  const k=19; c=656329; m=100000001;

  BEGIN
    rseed := ( k*rseed + c ) mod m;
    ranf := rseed/1e8
  END;


  PROCEDURE setrancut{cut : double};

  { Set cut for normal distribution }

  begin 
    normcut_ := cut;
  end;


  FUNCTION normranf {: double};

  { Generate random numbers with normal distribution (m=0, sigma=1)
    and cut normcut_ }

  const	maxiter = 100;

  var	i, j	: Integer;
	f, w	: double;

  begin
    j := 0;
    repeat
      j := succ(j);
      w := 0d0;
      for i:=1 to 12 do
	w := w + ranf;
      f := w - 6.0;
    until (abs(f) <= abs(normcut_)) or (j > maxiter) ;
    if j > maxiter then writeln('  *** fatal error in normranf');
    normranf := f;
  end;


  FUNCTION dtor {d : double}{ : double};
  
  BEGIN
    if pi = 0d0 then writeln('** pi not initialized');
    dtor := d * pi/180.0;
  END;


  FUNCTION sign{x : double}{ : integer};

  BEGIN
    if x >= 0d0 then
      sign := 1
    else
      sign := -1;
  END;


  Function GetAngle{x, y : double}{ : double};

  { Get the angle phi from x=cos(phi), y=sin(phi) where -pi <= phi <= pi }

  var	z	: double;

  begin
    if pi = 0d0 then writeln('** pi not initialized');
    if x <> 0 then
      z := ArcTan(y/x)
    else
      z := sign(y)*pi/2;
    if x < 0 then
    begin
      if y >= 0 then
	z := z + pi
      else
	z := z - pi;
    end;
    GetAngle := z;
  end;


  PROCEDURE UnitMat{n : integer; VAR a : matrix}; 

  var	i,j	:integer;

  begin
    for i:=1 to n  do
      for j:=1 to n do
	if i = j then
          a[i,j] := 1
	else
	  a[i,j] := 0d0;
  end;


  PROCEDURE CopyVec{n : integer; VAR a, b : vector};

  VAR	i	: integer; 

  BEGIN
    FOR i:=1 TO n DO 
      b[i] := a[i];
  END;


  PROCEDURE CopyMat{n : integer; VAR a, b : matrix};

  VAR	i	: integer;

  BEGIN
    FOR i:=1 TO n DO 
      CopyVec(n, a[i], b[i]);
  END;


  PROCEDURE AddVec{n : integer; VAR a, b : vector};

  VAR	i	: integer;

  BEGIN
    FOR i:=1 TO n DO 
      b[i] := a[i] + b[i];
  END;


  PROCEDURE SubVec{n : integer; VAR a, b : vector};

  VAR	i	: integer;

  BEGIN
    FOR i:=1 TO n DO 
      b[i] := b[i] - a[i];
  END;


  PROCEDURE AddMat{n : integer; VAR a, b : matrix};

  VAR	i	: integer;

  BEGIN
    FOR i:=1 TO n DO
      AddVec(n, a[i], b[i]);
  END;


  PROCEDURE SubMat{n : integer; VAR a, b : matrix};

  VAR	i	: integer;

  BEGIN
    FOR i:=1 TO n DO
      SubVec(n, a[i], b[i]);
  END;


  PROCEDURE LinTrans{n : integer; VAR a : matrix; VAR x : vector};

    VAR	i, j	: integer;
	y	: vector;

    BEGIN
      FOR i:=1 TO n DO
      BEGIN
        y[i] := 0d0;
        FOR j:=1 TO n DO
          y[i] := y[i] + a[i, j] * x[j];
      END;
      CopyVec(n, y, x);
    END;


  PROCEDURE MulLMat{n : integer; VAR a, b : matrix};

  VAR	i, j, k	: integer;
	x 	: double;
	c	: matrix;

  BEGIN
    FOR i:=1 TO n DO
      FOR j:=1 TO n DO
      BEGIN
        x := 0d0;
        FOR k := 1 TO n DO 
          x := x + a[i, k] * b[k, j];
        c[i, j] := x;
      END;
    CopyMat(n, c, b);
  END;


  PROCEDURE MulRMat{n : integer; VAR a, b : matrix};

  VAR	i, j, k	: integer;
	x 	: double;
	c	: matrix;

  BEGIN
    FOR i:=1 TO n DO
      FOR j:=1 TO n DO
      BEGIN
        x := 0d0;
        FOR k := 1 TO n DO 
          x := x + a[i, k] * b[k, j];
        c[i, j] := x;
      END;
    CopyMat(n, c, a);
  END;


  FUNCTION TrMat{n : integer; VAR a : matrix}{ : double};

    VAR	i	: integer;
	x	: double;

    BEGIN
      x := 0d0;
      FOR i:=1 TO n DO
        x := x + a[i, i];
      TrMat := x;
    END;


  PROCEDURE TpMat{n : integer; VAR a : matrix};

    VAR	i, j	: integer;
	x	: double;

    BEGIN
      FOR i:=1 TO n DO
        FOR j:=1 TO i-1 DO
	begin
          x := a[i, j]; a[i, j] := a[j, i]; a[j, i] := x;
	end;
    END;


  FUNCTION DetMat{n : integer; VAR a : matrix}{ : double};

    var		j	: integer;
		d	: double;
		cross	: array [1..matdim] of boolean;


    FUNCTION GdetMat(n : integer) : double;

      VAR	k, sign	: integer;
		det	: double;

      BEGIN
        if n > 1 then
        begin
          det := 0d0;
	  if n mod 2 = 1 then
	    sign := 1
	  else
	    sign := -1;
          for k:=1 to matdim do
	  begin
 	    if not cross[k] then
	    begin
	      cross[k] := true;
	      det := det + sign*a[n, k]*GdetMat(n-1);
	      cross[k] := false;
	      sign := -sign;
	    end;
	  end;
	  GdetMat := det;
        end
        else
          for k:=1 to matdim do
            if not cross[k] then GdetMat := a[n, k];
      END;

    begin
      if n = 2 then
        DetMat := a[1, 1]*a[2, 2] - a[1, 2]*a[2, 1]
      else if n = 3 then
      begin
        d :=     a[1, 1]*(a[2, 2]*a[3, 3] - a[2, 3]*a[3, 2]);
        d := d + a[1, 2]*(a[2, 3]*a[3, 1] - a[2, 1]*a[3, 3]);
        d := d + a[1, 3]*(a[2, 1]*a[3, 2] - a[2, 2]*a[3, 1]);
        DetMat := d;
      end
      else
      begin
        for j:=1 to matdim do
	  if j <= n then
	    cross[j] := false
	  else
	    cross[j] := true;
        DetMat := GdetMat(n);
      end;
    end;


  function InvMat{n : integer; VAR a : matrix}{ : boolean};

  { Gauss-Jordan elimination with full pivoting,
    gaussj from Numerical Recipes }

    LABEL      1;

    TYPE	iv1	= ARRAY[1..matdim] OF integer;
		iv2	= ARRAY[1..matdim, 1..2] OF integer;
		v1	= ARRAY[1..matdim] OF double;

    VAR	i, j, k, l, l1, row, column	: integer;
	amax, t, determ, d		: double;
	b				: matrix;
	ipivot				: iv1;
	index				: iv2;
	pivot				: v1;


    PROCEDURE swap (VAR x, y : double);

      VAR      d : double;

      BEGIN
	d := x; x := y; y := d
      END;


    PROCEDURE interchange;

      VAR        l : integer;

    BEGIN
      IF row <> column THEN
      BEGIN
        determ := -determ;
        FOR l := 1 TO n DO 
          swap(a[row, l], a[column, l]);
      END
    END;  {interchange}

  BEGIN {matinv}
    if n = 2 then
    begin
      d := detMat(2, a);
      IF d <> 0d0 THEN
      BEGIN
        InvMat := true;
        b[1, 1] :=  a[2, 2]/d;
        b[1, 2] := -a[1, 2]/d;
        b[2, 1] := -a[2, 1]/d;
        b[2, 2] :=  a[1, 1]/d;
	CopyMat(n, b, a);
      END
      ELSE
        InvMat := false;
    end
    else
    begin
      determ := 1.0;
      FOR j := 1 TO n DO ipivot[j] := 0;
      i := 1;
      WHILE (i <= n) AND (determ <> 0d0) DO
      BEGIN
        amax := 0d0;
        FOR j := 1 TO n DO
          IF ipivot[j] <> 1 THEN
            FOR k := 1 TO n DO
            BEGIN
              IF ipivot[k] > 1 THEN GOTO 1;
              IF ipivot[k] < 1 THEN
              BEGIN
                IF abs(amax) < abs(a[j, k]) THEN
                BEGIN
		  row := j; column := k; amax := a[j, k]
	        END
              END
	    END;
            IF amax = 0d0 THEN
            begin
	      InvMat := false;
	      determ := 0d0;
	    end
            ELSE
            BEGIN
	      InvMat := true;
              ipivot[column] := ipivot[column] + 1;
              interchange;
              index[i, 1] := row; index[i, 2] := column;
              pivot[i] := a[column, column];
              determ := determ * pivot[i];
              a[column, column] := 1.0;
              FOR l := 1 TO n DO 
                a[column, l] := a[column, l] / pivot[i];
              FOR l1 := 1 TO n DO
              BEGIN
                IF l1 <> column THEN
                BEGIN
                  t := a[l1, column];
                  a[l1, column] := 0d0;
                  FOR l := 1 TO n DO 
                    a[l1, l] := a[l1, l] - a[column, l] * t;
                END
              END
            END; {else }
            i := i + 1;
          END; {while}
          IF determ <> 0d0 THEN
            FOR i := 1 TO n DO
            BEGIN
              l := n + 1 - i;
              IF index[l, 1] <> index[l, 2] THEN
              BEGIN
		row := index[l, 1]; column := index[l, 2];
                FOR k := 1 TO n DO
		  swap(a[k, row], a[k, column])
              END
           END;
    end;
1 :
  END;


  procedure prtmat{n : integer; var a : matrix};

  var	i, j	: integer;

  begin
    writeln('matrix:');
    for i:=1 to n do
    begin
      for j:=1 to n do
        write(a[i, j]);
      writeln;
    end;
  end;

  end.
